app/Http/Controllers/Auth/LoginController.php

```diff
@@ -3,37 +3,140 @@
 namespace App\Http\Controllers\Auth;
 
 use App\Http\Controllers\Controller;
-use Illuminate\Foundation\Auth\AuthenticatesUsers;
+use App\Libraries\SSO\VATSIMSSO;
+use App\User;
+use Carbon\Carbon;
+use Exception;
+use Illuminate\Auth\AuthenticationException;
+use Illuminate\Http\Request;
+use Illuminate\Support\Facades\Auth;
+use Illuminate\Support\Facades\Input;
+use Illuminate\Support\Facades\Session;
 
 class LoginController extends Controller
 {
     /*
-    |--------------------------------------------------------------------------
-    | Login Controller
-    |--------------------------------------------------------------------------
-    |
-    | This controller handles authenticating users for the application and
-    | redirecting them to your home screen. The controller uses a trait
-    | to conveniently provide its functionality to your applications.
-    |
-    */
-
-    use AuthenticatesUsers;
-
-    /**
-     * Where to redirect users after login.
-     *
-     * @var string
+     * Login Lifecycle:
+     *  1) Anonymous user lands at /login
+     *  2) User is redirected to VATSIM SSO system
+     *  3) User returns from VATSIM SSO system, with authentication token, which is then validated
+     *  4) If valid, user is checked to see if they require Secondary Authentication. If not, they are logged in.
+     *  5) Secondary Authentication users are then logged in once completed successfully.
      */
-    protected $redirectTo = '/home';
 
-    /**
-     * Create a new controller instance.
-     *
-     * @return void
-     */
     public function __construct()
     {
         $this->middleware('guest')->except('logout');
     }
+
+    public function logout()
+    {
+        Auth::logout();
+        Auth::guard('partial_web')->logout();
+        return redirect('/');
+    }
+
+    /*
+     * Step 1: Redirect to VATSIM.NET SSO
+     */
+    public function loginWithVatsimSSO()
+    {
+		// Check we have nessesary information
+		if(!VATSIMSSO::isEnabled()){
+			return back()->with('error', 'VATSIM SSO Authentication is not currently available');
+		}
+		
+		
+        if(Auth::guard('partial_web')->check()){
+            return redirect()->route('login.secondary');
+        }
+
+        $sso = new VATSIMSSO();
+        return $sso->login(url('/login/sso/verify'), function ($key, $secret, $url) {
+            Session::put('vatsimauth', compact('key', 'secret'));
+            return redirect($url);
+        }, function ($error) {
+            throw new AuthenticationException('Could not authenticate: ' . $error['message']);
+        });
+    }
+
+    /*
+     * Step 2: Verify SSO Login after redirect
+     */
+    public function verifySSOLogin()
+    {
+        $sso = new VATSIMSSO();
+        $session = Session::get('vatsimauth');
+
+        return $sso->validate(
+            $session['key'],
+            $session['secret'],
+            Input::get('oauth_verifier'),
```

Comment:
Do want some validation on this input, check it at least exists? From a maintenance perspective, it would be better to get a 400 Bad Request if this is missing, rather than let the underlying SSO library throw an exception or return something ambiguous.